### PR TITLE
support multiple exit codes based on what went wrong/right

### LINKED ISSUE #1131: When `preflight` results in FAILED, it exits with code 0
**Bug Description**

When I run a preflight test (via `kubectl preflight` <some url>), the output might be, for example:

```
$ kubectl preflight --interactive=false ./preflight.yaml
name: cluster-resources    status: running         completed: 0    total: 2
name: cluster-resources    status: completed       completed: 1    total: 2
name: cluster-info         status: running         completed: 1    total: 2
name: cluster-info         status: completed       completed: 2    total: 2

   --- PASS Required Kubernetes Version
      --- Your cluster meets the recommended and required versions of Kubernetes.
   --- FAIL: Must have at least 3 nodes in the cluster
      --- This application requires at least 3 nodes
--- FAIL   preflight-sample
FAILED

$ echo $?
0
```

**Expected Behavior**

If I use this in a shell script or any other wrapper that uses preflights as a test to determine if the cluster is up to snuff or not, I need to grep for the output rather than simply use the exit code like most folks would be default.

I would expect the exit code for failed tests to be non-zero, though possibly different if there's an error actually running the preflights as opposed to preflights failing.

**Steps To Reproduce**

Preflight 0.62.1, Linux/Intel, k3d cluster with one node.
Very simple preflight check:

```yaml
apiVersion: troubleshoot.sh/v1beta2
kind: Preflight
metadata:
  name: preflight-sample
spec:
  analyzers:
    - clusterVersion:
        outcomes:
          - fail:
              when: "< 1.19.0"
              message: The application requires at least Kubernetes 1.19.0, and recommends 1.24.0.
              uri: https://kubernetes.io
          - warn:
              when: "< 1.23.0"
              message: Your cluster meets the minimum version of Kubernetes, but we recommend you update to 1.24.0 or later.
              uri: https://kubernetes.io
          - pass:
              message: Your cluster meets the recommended and required versions of Kubernetes.
    - nodeResources:
        checkName: Must have at least 3 nodes in the cluster
        outcomes:
          - fail:
              when: "count() < 3"
              message: This application requires at least 3 nodes
          - pass:
              message: This cluster has enough nodes.
```

From what I can see, [the preflight package return](https://github.com/replicatedhq/troubleshoot/blob/6493064e4599adb2a4f1255d90052872e73f5d8c/pkg/preflight/run.go#L265) doesn't do any check to actually return a boolean for the result, just the text of the result itself.

### PR DESCRIPTION
**Pull request by CpuID**
## Description, Motivation and Context

support multiple exit codes based on what went wrong/right

main motivator initially: failing helm preflights in an elegant way

0 = all passed, 3 = at least one failure, 4 = no failures but at least 1 warn

1 as a catch all, 2 for invalid input etc

resolves https://github.com/replicatedhq/troubleshoot/issues/1131

## Checklist

- [x] New and existing tests pass locally with introduced changes.
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] The commit message(s) are informative and highlight any breaking changes
- [x] Any documentation required has been added/updated. For changes to https://troubleshoot.sh/ create a PR [here](https://github.com/replicatedhq/troubleshoot.sh/pulls)

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by CpuID** [comment]
TODO: docs

update: done https://github.com/replicatedhq/troubleshoot.sh/pull/489

**Comment by CpuID** [comment]
TODO: fix what I did with `cli.RootCmd`, to still propagate the exit code and allow all tests to pass

**Comment by CpuID** [comment]
docs PR - https://github.com/replicatedhq/troubleshoot.sh/pull/489

**Comment by CpuID** [comment]
Ready for review 👍

Tests pass and docs are done https://github.com/replicatedhq/troubleshoot.sh/pull/489

File: pkg/preflight/run.go [hunk 1]
```diff
@@ -56,69 +58,74 @@ func RunPreflights(interactive bool, output string, format string, args []string
 	var uploadResultSpecs []*troubleshootv1beta2.Preflight
 	var err error
 
+	catchAllExitCode := 1
+	specIssuesExitCode := 2
```
Comment (DexterYan) [inline pkg/preflight/run.go]: Looks good. Maybe for those two, we can put to pkg/constants/constants.go

**Review by emosbaugh — COMMENTED** [review]
I feel like a more go-like pattern for exiting with different codes would be to define an interface

```go
type ExitError interface {
	Error() string
	ExitStatus() int
}

func main() {
	err := RootCmd().Execute()
	fmt.Fprintln(os.Stderr, err.Error())
	switch err := errors.Unwrap(err).(type) {
	case ExitError:
		os.Exit(err.ExitStatus())
	default:
		os.Exit(1)
	}
}
```

**Comment by CpuID** [comment]
> I feel like a more go-like pattern for exiting with different codes would be to define an interface

I actually considered something like this, and even went to the extent of a custom error type (struct) that implements the `error` interface with extra methods and properties, but no matter how I tried, the `RunE` function in cobra insists on nothing else but a vanilla `error` type being returned which is rather annoying :( eg. like https://pkg.go.dev/errors#example-package

The idea of using `errors.Unwrap()` is a good one though... as an alternative to that delimiter approach, just to try and split up the exit code vs the error itself. It wasn't quite 100% what I was looking for though...

I did almost have some success again today with a custom error struct (to the point I got it to build), but it was still just a generic error being returned not my custom type.

If I change the `RunE` response param to be my custom type, I end up with something like this which is the biggest issue:

```
cmd/preflight/cli/root.go:85:9: cannot use func(cmd *cobra.Command, args []string) ExitCodeAndError {…} (value of type func(cmd *cobra.Command, args []string) ExitCodeAndError) as type func(cmd *cobra.Command, args []string) error in struct literal
```

I think for now in light of wanting to get this merged ASAP for use, we should retain the delimiter approach and as a future improvement try refactor things here.

**Comment by CpuID** [comment]
@xavpaice also mentioned this example using `Run` instead of `RunE` - https://github.com/k8sgpt-ai/k8sgpt/blob/d8357ceb949e04d9dd21276a1d1dfcb60010c37a/cmd/analyze/analyze.go#L44

I've tried this approach also - the downside is it'll never run our `PostRun` hooks if we exit within `Run`, and we finish up our profiling logic in there right now... so I didn't bother going down this path (I'd like to keep `PostRun` as a usable thing)

**Comment by emosbaugh** [comment]
Apologies for not testing the code in my suggestion. This was more difficult than I had thought. Ive created a quick prototype for returning errors with exit codes [here](https://github.com/replicatedhq/troubleshoot/pull/1143).

**Comment by CpuID** [comment]
> Apologies for not testing the code in my suggestion. This was more difficult than I had thought. Ive created a quick prototype for returning errors with exit codes [here](https://github.com/replicatedhq/troubleshoot/pull/1143).

no worries thanks @emosbaugh :)

File: cmd/preflight/cli/root.go [hunk 2]
```diff
@@ -75,9 +79,20 @@ that a cluster meets the requirements to run an application.`,
...
+	if err != nil {
+		var exitErr types.ExitError
+		if errors.As(err, &exitErr) {
+			if len(exitErr.Error()) > 0 {
+				cmd.PrintErrln("Error:", err.Error())
+			}
+			os.Exit(exitErr.ExitStatus())
+		}
 		os.Exit(1)
 	}
```
Comment (emosbaugh) [inline cmd/preflight/cli/root.go]: you will need to print the error regardless as you are silencing errors in the cmd

```suggestion
	if err != nil {
		cmd.PrintErrln("Error:", err.Error())
		var exitErr types.ExitError
		if errors.As(err, &exitErr) {
			os.Exit(exitErr.ExitStatus())
		}
		os.Exit(1)
	}
```

File: cmd/preflight/cli/root.go [hunk 3]
```diff
@@ -75,9 +79,20 @@ that a cluster meets the requirements to run an application.`,
...
+	os.Exit(1)
```
Comment (emosbaugh) [inline cmd/preflight/cli/root.go]: you should not exit with an error if there is no error

```suggestion
```

**Review by emosbaugh — CHANGES_REQUESTED** [review]

**Comment by CpuID** [comment]
@emosbaugh the downside to always printing an error is we end up with the below, you get `Error:` with nothing after it at the end...

```
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ bin/preflight --interactive=false 1131-preflight-invalid.yaml                                                                                         14:38:17
name: cluster-resources    status: running         completed: 0    total: 2   
name: cluster-resources    status: completed       completed: 1    total: 2   
name: cluster-info         status: running         completed: 1    total: 2   
name: cluster-info         status: completed       completed: 2    total: 2   

   --- FAIL: nonexistent analyzer
      --- Analyzer not found
--- FAIL   preflight-sample
FAILED
Error: 
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ echo $?                                                                                                                                               10:23:03
3
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ bin/preflight --interactive=false 1131-preflight-pass.yaml                                                                                            10:23:05
name: cluster-resources    status: running         completed: 0    total: 2   
name: cluster-resources    status: completed       completed: 1    total: 2   
name: cluster-info         status: running         completed: 1    total: 2   
name: cluster-info         status: completed       completed: 2    total: 2   

   --- PASS Required Kubernetes Version
      --- Your cluster meets the recommended and required versions of Kubernetes.
--- PASS   preflight-sample
PASS
Error: 
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ echo $?                                                                                                                                               10:23:10
0
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ bin/preflight --interactive=false 1131-preflight-fail.yaml                                                                                            10:23:12
name: cluster-resources    status: running         completed: 0    total: 2   
name: cluster-resources    status: completed       completed: 1    total: 2   
name: cluster-info         status: running         completed: 1    total: 2   
name: cluster-info         status: completed       completed: 2    total: 2   

   --- PASS Required Kubernetes Version
      --- Your cluster meets the recommended and required versions of Kubernetes.
   --- FAIL: Must have at least 3 nodes in the cluster
      --- This application requires at least 3 nodes
--- FAIL   preflight-sample
FAILED
Error: 
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ echo $?                                                                                                                                               10:23:17
3
```

File: pkg/preflight/run.go [hunk 4]
```diff
@@ -255,14 +260,40 @@ func RunPreflights(interactive bool, output string, format string, args []string
...
+
 	if interactive {
-		if len(analyzeResults) == 0 {
-			return errors.New("no data has been collected")
+		err = showInteractiveResults(preflightSpecName, output, analyzeResults)
+	}
+
+	err = showTextResults(format, preflightSpecName, output, analyzeResults)
+
+	return types.NewExitCodeError(exitCode, err)
```
Comment (emosbaugh) [inline pkg/preflight/run.go]: You should not return an error if there is no error (exit code is 0)

**Review by emosbaugh — CHANGES_REQUESTED** [review]

**Comment by emosbaugh** [comment]
> @emosbaugh the downside to always printing an error is we end up with the below, you get `Error:` with nothing after it at the end...
> 
> ```
> ~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ bin/preflight --interactive=false 1131-preflight-invalid.yaml                                                                                         14:38:17
> name: cluster-resources    status: running         completed: 0    total: 2   
> name: cluster-resources    status: completed       completed: 1    total: 2   
> name: cluster-info         status: running         completed: 1    total: 2   
> name: cluster-info         status: completed       completed: 2    total: 2   
> 
>    --- FAIL: nonexistent analyzer
>       --- Analyzer not found
> --- FAIL   preflight-sample
> FAILED
> Error: 
> ~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ echo $?                                                                                                                                               10:23:03
> 3
> ~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ bin/preflight --interactive=false 1131-preflight-pass.yaml                                                                                            10:23:05
> name: cluster-resources    status: running         completed: 0    total: 2   
> name: cluster-resources    status: completed       completed: 1    total: 2   
> name: cluster-info         status: running         completed: 1    total: 2   
> name: cluster-info         status: completed       completed: 2    total: 2   
> 
>    --- PASS Required Kubernetes Version
>       --- Your cluster meets the recommended and required versions of Kubernetes.
> --- PASS   preflight-sample
> PASS
> Error: 
> ~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ echo $?                                                                                                                                               10:23:10
> 0
> ~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ bin/preflight --interactive=false 1131-preflight-fail.yaml                                                                                            10:23:12
> name: cluster-resources    status: running         completed: 0    total: 2   
> name: cluster-resources    status: completed       completed: 1    total: 2   
> name: cluster-info         status: running         completed: 1    total: 2   
> name: cluster-info         status: completed       completed: 2    total: 2   
> 
>    --- PASS Required Kubernetes Version
>       --- Your cluster meets the recommended and required versions of Kubernetes.
>    --- FAIL: Must have at least 3 nodes in the cluster
>       --- This application requires at least 3 nodes
> --- FAIL   preflight-sample
> FAILED
> Error: 
> ~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ echo $?                                                                                                                                               10:23:17
> 3
> ```

I overlooked this logic on my previous review. All errors should have a message. You should set the error message to something like "preflight checks failed with errors" or "preflight checks failed with warnings". If you dont want to output a message in this case I would suggest another error type PreflightCheckResultsError and have another case in InitAndExecute with errors.As where you do not print the error.

File: pkg/preflight/run.go (hunk 4 above)
Comment (CpuID) [reply pkg/preflight/run.go]: I think I've got a good enough way to do this now 👍

**Comment by CpuID** [comment]
OK I feel like this might be the best we're going to get... keen to get this merged and then improve on it later if we need to though :)

```
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ bin/preflight --interactive=false 1131-preflight-pass.yaml                                                                                 
name: cluster-resources    status: running         completed: 0    total: 2   
name: cluster-resources    status: completed       completed: 1    total: 2   
name: cluster-info         status: running         completed: 1    total: 2   
name: cluster-info         status: completed       completed: 2    total: 2   

   --- PASS Required Kubernetes Version
      --- Your cluster meets the recommended and required versions of Kubernetes.
--- PASS   preflight-sample
PASS
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ echo $?                                                                                                                                               
0
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ bin/preflight --interactive=false 1131-preflight-fail.yaml                                                                                            
name: cluster-resources    status: running         completed: 0    total: 2   
name: cluster-resources    status: completed       completed: 1    total: 2   
name: cluster-info         status: running         completed: 1    total: 2   
name: cluster-info         status: completed       completed: 2    total: 2   

   --- PASS Required Kubernetes Version
      --- Your cluster meets the recommended and required versions of Kubernetes.
   --- FAIL: Must have at least 3 nodes in the cluster
      --- This application requires at least 3 nodes
--- FAIL   preflight-sample
FAILED
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ echo $?                                                                                                                                               
3
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ bin/preflight --interactive=false 1131-preflight-invalid.yaml                                                                                         
name: cluster-resources    status: running         completed: 0    total: 2   
name: cluster-resources    status: completed       completed: 1    total: 2   
name: cluster-info         status: running         completed: 1    total: 2   
name: cluster-info         status: completed       completed: 2    total: 2   

   --- FAIL: nonexistent analyzer
      --- Analyzer not found
--- FAIL   preflight-sample
FAILED
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ echo $?                                                                                                                                               
3
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ # ^^ this ones hard to return a 2 for (instead of a 3), as its in the analyzer code itself. I didn't want to end up with a bunch of extra scope creep, left it as a 3
quote> 
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ bin/preflight --interactive=false 1131-preflight-pass.yaml --format asdf                                                                        
name: cluster-resources    status: running         completed: 0    total: 2   
name: cluster-resources    status: completed       completed: 1    total: 2   
name: cluster-info         status: running         completed: 1    total: 2   
name: cluster-info         status: completed       completed: 2    total: 2   
Error: unknown output format: "asdf"
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ echo $?                                                                                                                                               
2
```

File: cmd/preflight/cli/root.go [hunk 5]
```diff
@@ -83,13 +83,24 @@ func InitAndExecute() {
...
 		if errors.As(err, &exitErr) {
 			os.Exit(exitErr.ExitStatus())
 		}
+
+		// Fallback, should almost never be used (the above Exit() should handle almost all situations
 		os.Exit(1)
 	}
+
+	// Everything went fine
+	os.Exit(0)
```
Comment (emosbaugh) [inline cmd/preflight/cli/root.go]: I would prefer we do not exit here.

```suggestion

	// Everything went fine
	os.Exit(0)
```

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: cmd/preflight/cli/root.go (hunk 5 above)
Comment (CpuID) [reply cmd/preflight/cli/root.go]: fixed

**Comment by CpuID** [comment]
```
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ bin/preflight --interactive=false 1131-preflight-fail.yaml --format json                                                                              07:50:40
name: cluster-resources    status: running         completed: 0    total: 2   
name: cluster-resources    status: completed       completed: 1    total: 2   
name: cluster-info         status: running         completed: 1    total: 2   
name: cluster-info         status: completed       completed: 2    total: 2   
{
  "pass": [
    {
      "title": "Required Kubernetes Version",
      "message": "Your cluster meets the recommended and required versions of Kubernetes."
    }
  ],
  "fail": [
    {
      "title": "Must have at least 3 nodes in the cluster",
      "message": "This application requires at least 3 nodes"
    }
  ]
}
```

**Comment by CpuID** [comment]
```
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ bin/preflight --interactive=false 1131-preflight-fail.yaml >/dev/null                                                                                 07:51:38
name: cluster-resources    status: running         completed: 0    total: 2   
name: cluster-resources    status: completed       completed: 1    total: 2   
name: cluster-info         status: running         completed: 1    total: 2   
name: cluster-info         status: completed       completed: 2    total: 2   
~/git/replicatedhq/troubleshoot cpuid-1131 *6 ?3 ❯ bin/preflight --interactive=false 1131-preflight-fail.yaml 2>/dev/null                                                                                07:51:43

   --- PASS Required Kubernetes Version
      --- Your cluster meets the recommended and required versions of Kubernetes.
   --- FAIL: Must have at least 3 nodes in the cluster
      --- This application requires at least 3 nodes
--- FAIL   preflight-sample
FAILED
```

File: pkg/preflight/run.go [hunk 6]
```diff
@@ -255,14 +260,50 @@ func RunPreflights(interactive bool, output string, format string, args []string
...
+		return types.NewExitCodeError(constants.EXIT_CODE_CATCH_ALL, errors.New("no data has been collected"))
+	}
+
 	if interactive {
-		if len(analyzeResults) == 0 {
-			return errors.New("no data has been collected")
+		err = showInteractiveResults(preflightSpecName, output, analyzeResults)
+	}
+
+	err = showTextResults(format, preflightSpecName, output, analyzeResults)
```
Comment (emosbaugh) [inline pkg/preflight/run.go]:
```suggestion
		err = showInteractiveResults(preflightSpecName, output, analyzeResults)
	} else {
		err = showTextResults(format, preflightSpecName, output, analyzeResults)
	}
```

File: pkg/preflight/run.go [hunk 7]
```diff
@@ -255,14 +260,50 @@ func RunPreflights(interactive bool, output string, format string, args []string
...
+
+	var exitCode int
+	if err != nil {
+		// Always return a 2 here
+		exitCode = 2
+	} else {
+		exitCode = checkOutcomesToExitCode(analyzeResults)
+	}
```
Comment (emosbaugh) [inline pkg/preflight/run.go]: this feels like it should be a standard exit code here as it is not a spec issue

```suggestion
	if err != nil {
		return errors.Wrap(err, "failed to print preflight results")
	}

	exitCode := checkOutcomesToExitCode(analyzeResults)
```

File: pkg/preflight/run.go [hunk 8]
```diff
@@ -255,14 +260,50 @@ func RunPreflights(interactive bool, output string, format string, args []string
...
+
+	if exitCode == 0 {
+		return nil
+	}
+
+	return types.NewExitCodeError(exitCode, err)
```
Comment (emosbaugh) [inline pkg/preflight/run.go]:
```suggestion
	return types.NewExitCodeError(exitCode, errors.New("preflights failed with warnings or errors"))
```

**Review by emosbaugh — CHANGES_REQUESTED** [review]

**Review by emosbaugh — APPROVED** [review]

**Review by xavpaice — APPROVED** [review]

**Pull request merged**